contracts/YieldCrabLever.sol

```diff
@@ -0,0 +1,451 @@
+                ilkId, // [19:25]
+                bytes32(ink), // [25:57]
+                bytes32(art)
+            );
+
+            address join = address(ladle.joins(seriesId & ASSET_ID_MASK));
+
+            // Close:
+            // Series is not past maturity, borrow and move directly to collateral pool.
+            // We have a debt in terms of fyToken, but should pay back in base.
+            uint128 base = cauldron.debtToBase(seriesId, art.u128());
+            success = IERC3156FlashLender(join).flashLoan(
+                this, // Loan Receiver
+                address(IJoin(join).asset()), // Loan Token
+                base, // Loan Amount
+                data
+            );
+        } else {
+            IPool pool = IPool(ladle.pools(seriesId));
+            IMaturingToken fyToken = pool.fyToken();
+            // Repay:
+            // Series is not past maturity.
+            // Borrow to repay debt, move directly to the pool.
+            bytes memory data = bytes.concat(
+                bytes1(bytes1(uint8(uint256(Operation.REPAY)))), // [0:1]
+                seriesId, // [1:7]
+                vaultId, // [7:19]
+                ilkId, // [19:25]
+                bytes32(ink), // [25:57]
+                bytes32(art) // [57:89]
+            );
+            success = IERC3156FlashLender(address(fyToken)).flashLoan(
+                this, // Loan Receiver
+                address(fyToken), // Loan Token
+                art, // Loan Amount: borrow exactly the debt to repay.
+                data
+            );
+            require(IERC20(address(fyToken)).balanceOf(address(this)) == 0);
+        }
+        if (!success) revert FlashLoanFailure();
+        IERC20 token = IERC20(cauldron.assets(ilkId));
+        uint256 tokenBalance = token.balanceOf(address(this));
+        if (tokenBalance < minBaseOut) revert SlippageFailure();
+        // Transferring the leftover to the user
+        token.safeTransfer(msg.sender, tokenBalance);
+        // Give the vault back to the sender, just in case there is anything left
+        giver.give(vaultId, msg.sender);
+    }
+
+    /// @notice Callback for the flash loan.
+    /// @dev Used as a router to the correct function based
+    ///      on the operation present in the data.
+    function onFlashLoan(
+        address initiator,
+        address token, // The token, not checked as we check the lender address.
+        uint256 borrowAmount,
+        uint256 fee,
+        bytes calldata data
+    ) external override returns (bytes32 returnValue) {
+        returnValue = FLASH_LOAN_RETURN;
+
+        Operation status = Operation(uint256(uint8(data[0])));
+        bytes6 seriesId = bytes6(data[1:7]);
+        IPool pool = IPool(ladle.pools(seriesId));
+
+        // Test that the lender is either a fyToken contract or the join.
+        if (
+            msg.sender != address(pool.fyToken()) &&
+            msg.sender != address(ladle.joins(seriesId & ASSET_ID_MASK))
```

Comment:
```suggestion
            msg.sender != address(ladle.joins(cauldron.series(seriesId).baseId))
```

Reply:
You also receive `baseId` in `data`. It is ok to trust `data` here to identify the caller because it needs to be one of our joins or fyToken.